### PR TITLE
District 3

### PR DESCRIPTION
**Pull request by Yoel-Perman-Brilliant**

### CODE REVIEW

File: src/main/java/frc/robot/subsystems/ClimberWinch.java [hunk 1]
```diff
@@ -35,14 +42,30 @@ public static ClimberWinch getInstance() {
...
+//    }
 
     private ClimberWinch(CANSparkMax leftWinch, CANSparkMax rightWinch) {
         super(DOWN_SPEED, UP_SPEED, "climber winch", leftWinch, rightWinch);
-        leftWinch.setIdleMode(CANSparkMax.IdleMode.kBrake);
-        rightWinch.setIdleMode(CANSparkMax.IdleMode.kBrake);
-        encoder = rightWinch.getEncoder();
+        left = leftWinch;
+        right = rightWinch;
+        left.setIdleMode(CANSparkMax.IdleMode.kBrake);
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/subsystems/ClimberWinch.java:57]: use the internal `setIdleMode` method

File: src/main/java/frc/robot/commands/MoveIntakePlacerDown.java [hunk 2]
```diff
@@ -1,24 +1,41 @@
...
-        IntakePlacer intakePlacer = (IntakePlacer) subsystem;
-        intakePlacer.setServoAngle(IntakePlacer.SERVO_TARGET_ANGLE);
+        intakePlacer.setServoAngle(IntakePlacer.SERVO_TARGET_ANGLE.get());
+    }
+
+    @Override
+    public boolean isFinished() {
+        if (!ignoreLimit)
+            return intakePlacer.isDown();
+        else
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/commands/MoveIntakePlacerDown.java:33]: image

File: src/main/java/frc/robot/commands/MoveIntakePlacerDown.java [hunk 3]
```diff
@@ -1,24 +1,41 @@
...
+ */
+public class MoveIntakePlacerDown extends CommandBase {
 
-    public MoveIntakePlacerDown() {
-        super(IntakePlacer.getInstance(), IntakePlacer.MIN_SPEED);
+    private final IntakePlacer intakePlacer;
+    private final boolean ignoreLimit;
+
+    /**
+     * @param ignoreLimit represents whether the command should end once the latch has been opened, or wait until the
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/commands/MoveIntakePlacerDown.java:15]: remove `represents`

File: src/main/java/frc/robot/commands/MoveIntakePlacerDown.java [hunk 4]
```diff
@@ -1,24 +1,41 @@
 package frc.robot.commands;
 
-import com.spikes2212.command.genericsubsystem.commands.MoveGenericSubsystem;
+import edu.wpi.first.wpilibj2.command.CommandBase;
 import frc.robot.subsystems.IntakePlacer;
 
-public class MoveIntakePlacerDown extends MoveGenericSubsystem {
+/**
+ * Opens the intake placer's latch so the intake placer can move down.
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/commands/MoveIntakePlacerDown.java:7]: What does it do code-wise?

**Review by OfriRosenbaum — DISMISSED** [review]

File: src/main/java/frc/robot/commands/autonomous/YeetAndRetreat.java [hunk 5]
```diff
@@ -10,11 +11,13 @@
     public static final double RELEASE_DURATION = 1.0;
     public static final double DRIVE_SPEED = 0.7;
     public static final double DRIVE_ROTATE = 0;
-    public static final double DRIVE_DURATION = 1;
+    public static final double DRIVE_DURATION = 1.2212;
+    public static final double WAIT_DURATION = 3;
```
Comment (Spikes-2212) [inline src/main/java/frc/robot/commands/autonomous/YeetAndRetreat.java:15]: explain `WAIT_DURATION`

File: src/main/java/frc/robot/subsystems/ClimberWinch.java [hunk 6]
```diff
@@ -35,14 +43,29 @@ public static ClimberWinch getInstance() {
...
-        leftWinch.setIdleMode(CANSparkMax.IdleMode.kBrake);
-        rightWinch.setIdleMode(CANSparkMax.IdleMode.kBrake);
-        encoder = rightWinch.getEncoder();
+        left = leftWinch;
+        right = rightWinch;
+        setIdleMode(IdleMode.kBrake);
+        encoder = right.getEncoder();
     }
 
+//    private ClimberWinch(CANSparkMax rightWinch) {
```
Comment (Spikes-2212) [inline src/main/java/frc/robot/subsystems/ClimberWinch.java:62]: remove

File: src/main/java/frc/robot/subsystems/ClimberWinch.java [hunk 7]
```diff
@@ -35,14 +43,29 @@ public static ClimberWinch getInstance() {
         }
         return instance;
     }
+//    public static ClimberWinch getInstance() {
```
Comment (Spikes-2212) [inline src/main/java/frc/robot/subsystems/ClimberWinch.java:46]: remove

File: src/main/java/frc/robot/subsystems/ClimberWinch.java [hunk 8]
```diff
@@ -18,15 +19,22 @@
...
+    //@todo calibrate needed values
     public final Supplier<Double> ENCODER_UP_LIMIT =
-            rootNamespace.addConstantDouble("encoder up limit", 90);
+            rootNamespace.addConstantDouble("encoder up limit", 100);
     public final Supplier<Double> ENCODER_DOWN_LIMIT =
-            rootNamespace.addConstantDouble("encoder down limit", -30);
-
-    private static ClimberWinch instance;
+            rootNamespace.addConstantDouble("encoder down limit", 1);
+    public final Supplier<Double> ENCODER_STATIC_MEET_BAR_POSITION =
```
Comment (Spikes-2212) [inline src/main/java/frc/robot/subsystems/ClimberWinch.java:27]: remove

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 9]
```diff
@@ -49,12 +49,12 @@
     private final Supplier<Double> waitTimeGyro = gyroPIDNamespace.addConstantDouble("wait time", 0.1);
     private final PIDSettings pidSettingsGyro;
 
-    private final Supplier<Double> kPEncoders = encodersPIDNamespace.addConstantDouble("kP", 0);
-    private final Supplier<Double> kIEncoders = encodersPIDNamespace.addConstantDouble("kI", 0);
-    private final Supplier<Double> kDEncoders = encodersPIDNamespace.addConstantDouble("kD", 0);
-    private final Supplier<Double> toleranceEncoders = encodersPIDNamespace.addConstantDouble("tolerance", 0);
-    private final Supplier<Double> waitTimeEncoders = encodersPIDNamespace.addConstantDouble("wait time", 0);
-    private final PIDSettings pidSettingsEncoders;
+//    private final Supplier<Double> kPEncoders = encodersPIDNamespace.addConstantDouble("kP", 0);
```
Comment (Spikes-2212) [inline src/main/java/frc/robot/subsystems/Drivetrain.java:52]: remove

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 10]
```diff
@@ -25,7 +25,7 @@
     private static Drivetrain instance;
 
     private final Namespace encoderNamespace = rootNamespace.addChild("encoders");
-    private final Namespace encodersPIDNamespace = encoderNamespace.addChild("encoders pid");
+//    private final Namespace encodersPIDNamespace = encoderNamespace.addChild("encoders pid");
```
Comment (Spikes-2212) [inline src/main/java/frc/robot/subsystems/Drivetrain.java:28]: remove

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 11]
```diff
@@ -102,8 +102,8 @@ private Drivetrain(MotorControllerGroup leftMotors, BustedMotorControllerGroup r
         this.rightEncoder.setDistancePerPulse(DISTANCE_PER_PULSE);
         this.pidSettingsGyro = new PIDSettings(this.kPGyro, this.kIGyro, this.kDGyro, this.toleranceGyro,
                 this.waitTimeGyro);
-        this.pidSettingsEncoders = new PIDSettings(this.kPEncoders, this.kIEncoders, this.kDEncoders,
-                this.toleranceEncoders, this.waitTimeEncoders);
+//        this.pidSettingsEncoders = new PIDSettings(this.kPEncoders, this.kIEncoders, this.kDEncoders,
```
Comment (Spikes-2212) [inline src/main/java/frc/robot/subsystems/Drivetrain.java:105]: remove

File: src/main/java/frc/robot/subsystems/IntakePlacer.java [hunk 12]
```diff
@@ -15,10 +19,13 @@
  */
 public class IntakePlacer extends MotoredGenericSubsystem {
 
-    public static final double MAX_SPEED = 0.5;
+    public static final double MAX_SPEED = 0.6;
     public static final double MIN_SPEED = -0.1;
-    public static final double SERVO_START_ANGLE = 110;
-    public static final double SERVO_TARGET_ANGLE = 0;
+    private static final RootNamespace servoNamespace = new RootNamespace("servo");
+    public static final Supplier<Double> SERVO_START_ANGLE =
```
Comment (Spikes-2212) [inline src/main/java/frc/robot/subsystems/IntakePlacer.java:25]: change to 0

File: src/main/java/frc/robot/subsystems/IntakePlacer.java [hunk 13]
```diff
@@ -15,10 +19,13 @@
...
+            servoNamespace.addConstantDouble("start angle", 120);
+    public static final Supplier<Double> SERVO_TARGET_ANGLE =
```
Comment (Spikes-2212) [inline src/main/java/frc/robot/subsystems/IntakePlacer.java:27]: change to 90

File: src/main/java/frc/robot/commands/MoveIntakePlacerDown.java [hunk 14]
```diff
@@ -1,24 +1,39 @@
...
-import com.spikes2212.command.genericsubsystem.commands.MoveGenericSubsystem;
+import edu.wpi.first.wpilibj2.command.CommandBase;
 import frc.robot.subsystems.IntakePlacer;
 
-public class MoveIntakePlacerDown extends MoveGenericSubsystem {
+    /**
+ * Opens the intake placer's latch so the intake placer can move down, by rotating a servo that is
+ * connected to the latch.
+ */
+public class MoveIntakePlacerDown extends CommandBase {
```
Comment (Spikes-2212) [inline src/main/java/frc/robot/commands/MoveIntakePlacerDown.java:10]: change according to what I said to you irl

File: src/main/java/frc/robot/subsystems/IntakeToTransfer.java [hunk 15]
```diff
@@ -18,14 +19,14 @@
...
+        if (instance == null)
             instance = new IntakeToTransfer(new WPI_VictorSPX(RobotMap.CAN.INTAKE_TO_TRANSFER_VICTOR));
-        }
+
         return instance;
     }
 
-    private IntakeToTransfer(WPI_VictorSPX victor) {
-        super("intake to transfer", victor);
+    private IntakeToTransfer(MotorController motor) {
```
Comment (Spikes-2212) [inline src/main/java/frc/robot/subsystems/IntakeToTransfer.java:28]: why not victor?

File: src/main/java/frc/robot/commands/MoveIntakePlacerUp.java [hunk 16]
```diff
@@ -0,0 +1,32 @@
+package frc.robot.commands;
+
+import com.spikes2212.command.genericsubsystem.commands.MoveGenericSubsystem;
+import edu.wpi.first.wpilibj.Timer;
+import frc.robot.subsystems.IntakePlacer;
+
+/**
+ * Moves the IntakePlacer up until: <br>
+ * 1. The upper limit is hit. <br>
```
[the remark this answers is not in the document]
Comment (Yoel-Perman-Brilliant) [reply src/main/java/frc/robot/commands/MoveIntakePlacerUp.java:9]: why why?

File: src/main/java/frc/robot/subsystems/IntakeToTransfer.java (hunk 15 above)
Comment (Spikes-2212) [reply src/main/java/frc/robot/subsystems/IntakeToTransfer.java:28]: no

**Review by Spikes-2212 — APPROVED** [review]

**Pull request merged**